### PR TITLE
add IZ safeAccess util and refactor several && chains with it

### PR DESCRIPTION
**Pull request by BBasile**

### CODE REVIEW

File: src/dscanner/analysis/static_if_else.d [hunk 1]
```diff
@@ -47,17 +48,7 @@ class StaticIfElse : BaseAnalyzer
...
-			if (cc.falseStatement.statement.statementNoCaseNoDefault)
-			{
-				if (cc.falseStatement.statement.statementNoCaseNoDefault.ifStatement)
-				{
-					return cc.falseStatement.statement.statementNoCaseNoDefault.ifStatement;
-				}
-			}
-		}
-		return null;
+		return safeAccess(cc).falseStatement.statement.statementNoCaseNoDefault.ifStatement;
```
Comment (BBasile) [inline src/dscanner/analysis/static_if_else.d]: This change is :1st_place_medal:

File: src/dscanner/analysis/mismatched_args.d [hunk 2]
```diff
@@ -126,16 +127,16 @@ final class ArgVisitor : ASTVisitor
...
-			return;
-		if (unary.primaryExpression.identifierOrTemplateInstance is null)
-			return;
-		if (unary.primaryExpression.identifierOrTemplateInstance.identifier == tok!"")
-			return;
-		immutable t = unary.primaryExpression.identifierOrTemplateInstance.identifier;
-		lines ~= t.line;
-		columns ~= t.column;
-		args ~= internString(t.text);
+		if (const IdentifierOrTemplateInstance iot = safeAccess(unary)
```
Comment (wilzbach) [inline src/dscanner/analysis/mismatched_args.d]: Wouldn't `const iot` work too?

File: src/dscanner/analysis/static_if_else.d (hunk 1 above)
Comment (wilzbach) [reply src/dscanner/analysis/static_if_else.d]: Another bonus point for UFCS: `cc.safeAccess.falseStatement....` ;-)

File: src/dscanner/utils.d [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+				static if (is(R == class))
+				{
+					return (m is null)
+						? SafeAccess!R(null)
+						: SafeAccess!R(__traits(getMember, m, member)(a));
+				}
+				else
+				{
+                    if (m)
+					    __traits(getMember, m, member)(a);
```
Comment (wilzbach) [inline src/dscanner/utils.d]: weird indentation style

File: src/dscanner/analysis/mismatched_args.d (hunk 2 above)
Comment (BBasile) [reply src/dscanner/analysis/mismatched_args.d]: Yes. It can be removed unless DCD completion is needed.

Comment (BBasile) [reply src/dscanner/analysis/mismatched_args.d]: Wrong, the Type is actually required, otherwise the internal wrapper is returned.

Comment (BBasile) [reply src/dscanner/analysis/mismatched_args.d]: with `unwrap()` inference works

Comment (wilzbach) [reply src/dscanner/analysis/mismatched_args.d]: So why not use `unwrap` then?

Comment (BBasile) [reply src/dscanner/analysis/mismatched_args.d]: That's more o matter of style. I tend to use the type when i know it. Also better for DCD.

File: src/dscanner/utils.d [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		{
+			import std.traits : ReturnType, Parameters, isFunction;
+			static if (isFunction!T)
+			{
+				// otherwise there's a missing return statement.
+				alias R = ReturnType!T;
+				static if (!is(R == void) &&
+					!(is(R == class) && Parameters!T.length == 0))
+						pragma(msg, __FILE__ ~ "(" ~ __LINE__.stringof ~ "): error, " ~
+						"only `void function`s or `class` getters can be called without unwrap");
```
Comment (wilzbach) [inline src/dscanner/utils.d]: shouldn't this be an `static assert(0`, s.t. the compilation stops?
+ nit: use backticks with `unwrap`, s.t. DMD's colored highlighting works

Comment (wilzbach) [reply src/dscanner/utils.d]: Also why can't we return the result here?

Comment (BBasile) [reply src/dscanner/utils.d]: > shouldn't this be an static assert(0, s.t. the compilation stops?

the problem is that opDispatch gags the error messages so what happens exactly is that i let the error happens and use pragma to emit a pseudo error message.

**Pull request merged**